### ISSUE TITLE
Children reaping (waitpid) with external polling library

### ISSUE THREAD

**Issue by pparadzik** [warmcat/libwebsockets#950]
Hi,

When doing CGI calls with external polling library (uloop), libwebsockets doesn't reap the children if the polling library calls `waitpid` before. As a consequence of this, libwebsockets never sends `LWS_CALLBACK_CGI_TERMINATED` and closes connection after CGI is finished because `waitpid` in `lws_cgi_kill_terminated` never returns file descriptor of the terminated child. Currently I'm using `lws_service_fd` to inform libwebsockets about event that happend and it seems that libwebsockets always returns `LWS_CALLBACK_CGI` because it doesn't know that child is finished. Is there a way to handle this kind of situation?

**Comment by lws-team** [comment]
Hmm basically something else in your process also reaps child processes of the process and gets in the way of lws doing the job.  If I recall correctly lws also has two backup schemes for cgi children... a timeout on the cgi process, and a periodic poll of the child PIDs to make sure they are still there.  So despite this, lws should act reasonably, in terms of closing the logical cgi, if a bit late... is that what you experience?

**Comment by pparadzik** [comment]
Timeout works, lws periodically calls `lws_service_timeout_check` and closes connection after it happened (I forgot to mention that). Problem is that this takes too long if the CGI timeout value needs to be bigger, like 5 seconds. What happens is that some CGI's are not returning `content-length` or anything else to the client that would indicate size of the body, so client knows that it received everything only after lws closes connection which happens depending on the CGI timeout value. I'm not sure what do you mean by "periodic poll of the child PIDs".

Because I'm using external polling library, which is uloop, I'm relying on the `LWS_CALLBACK_ADD_POLL_FD`, `LWS_CALLBACK_DEL_POLL_FD` and similar callbacks. When CGI process is spawned, I receive `LWS_CALLBACK_ADD_POLL_FD`and add that file descriptor for polling to the uloop. After this CGI process finishes, uloop reads its return status and notifies me about it, after which I'm using `lws_service_fd` to notify lws. Is it possible to explicitly tell lws that child process is finished (someone else has  reaped it)? This way, I could notify lws when reaping happens and it could close connection associated with this CGI immediately.

**Comment by lws-team** [comment]
By periodic polling I meant calling this

https://github.com/warmcat/libwebsockets/blob/master/lib/libwebsockets.c#L2959

at intervals... notice it checks the exact child pids.

I think the problem here is uloop reaping all process children, not lws trying to manage his own child lifecycle.  TBH it's so specific a problem I'm not sure adding an api is the right way.

How about making your cgis do chunked and disable the uloop reaping business?  That sgould get you the result you want.

**Comment by pparadzik** [comment]
I don't have control over CGI's and it seems that uloop does not have an option to disable child reaping. I agree that problem is little to specific and was just wandering if there is some other option other then periodic polling and timeout which I'm aware of. Although, if such function existed, it seems that I would still not be able to tell it which file descriptor is related to the CGI that finished because I don't have that information in the `LWS_CALLBACK_ADD_POLL_FD` callback.

**Comment by lws-team** [comment]
Well the behaviour with "dumb" CGI is a separate issue.  For CGI where we can synchronously know they finished emitting (because we saw content-length, or it's chunked and we see the chunked EOF), and we know what's happened by reading the emitted data, lws does it perfect.  If the CGI is "dumb" and gives no sign that it finished emitting except, asynchronously, we find out the process ended some time later, we wait a bit to ensure we got all the stdout / stderr, and then do the CGI logical close.

My original point about the timeout is there is no huge problem here... it just waits a little longer IIUI and CGI timeout will clean up.

The fact uloop lacks an option to stop it killing your program doesn't make that part my problem.  The global process reaping thing is uloop's problem and you'll have to look there for a solution.
